flink-table-store-core/src/main/java/org/apache/flink/table/store/file/schema/SchemaEvolutionUtil.java

```diff
@@ -62,4 +72,193 @@ public static int[] createIndexMapping(
+     *
+     * @param tableFields the table fields
+     * @param dataFields the underlying data fields
+     * @param filters the filters
+     * @return the data filters
+     */
+    public static List<Predicate> createDataFilters(
+            List<DataField> tableFields, List<DataField> dataFields, List<Predicate> filters) {
+        if (filters == null) {
+            return null;
+        }
+
+        List<Predicate> dataFilters = new ArrayList<>(filters.size());
+        for (Predicate predicate : filters) {
+            dataFilters.add(createDataPredicate(tableFields, dataFields, predicate));
+        }
+        return dataFilters;
+    }
+
+    @Nullable
+    private static Predicate createDataPredicate(
+            List<DataField> tableFields, List<DataField> dataFields, Predicate predicate) {
+        if (predicate instanceof CompoundPredicate) {
+            CompoundPredicate compoundPredicate = (CompoundPredicate) predicate;
+            List<Predicate> children = compoundPredicate.children();
+            List<Predicate> dataChildren = new ArrayList<>(children.size());
+            for (Predicate child : children) {
+                Predicate dataPredicate = createDataPredicate(tableFields, dataFields, child);
+                if (dataPredicate != null) {
+                    dataChildren.add(dataPredicate);
+                }
+            }
+            return new CompoundPredicate(compoundPredicate.function(), dataChildren);
+        } else if (predicate instanceof LeafPredicate) {
+            LeafPredicate leafPredicate = (LeafPredicate) predicate;
+            List<DataField> predicateTableFields =
+                    tableFields.stream()
+                            .filter(f -> f.name().equals(leafPredicate.fieldName()))
+                            .collect(Collectors.toList());
+            if (predicateTableFields.size() != 1) {
+                throw new IllegalArgumentException(
+                        String.format("Find none or multiple fields %s", predicateTableFields));
+            }
+            DataField tableField = predicateTableFields.get(0);
+            List<DataField> predicateDataFields =
+                    dataFields.stream()
+                            .filter(f -> f.id() == tableField.id())
+                            .collect(Collectors.toList());
+            if (predicateDataFields.isEmpty()) {
+                return null;
+            } else if (predicateDataFields.size() > 1) {
+                throw new IllegalArgumentException(
+                        String.format("Find none or multiple fields %s", predicateTableFields));
+            }
+            DataField dataField = predicateDataFields.get(0);
+            return new LeafPredicate(
+                    leafPredicate.function(),
+                    leafPredicate.type(),
+                    dataFields.indexOf(dataField),
+                    dataField.name(),
+                    leafPredicate.literals());
```

Comment:
There are also troublesome things here. We may need to deal with type promotion (such as int ->bigint). I don't know whether orc will help me with this.

Reply:
We can add TODO here.